torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+        r.elem = elem
+        return r
+
+    def __repr__(self):
+        return f"CompositeCompliantTensor({self.elem})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            return e.elem if isinstance(e, CompositeCompliantTensor) else e
+
+        def wrap(e):
+            return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
+
+        if func.__name__ in ('set_', 'resize_'):
+            raise RuntimeError(
+                f"{func.__name__} is not allowed to be called inside of "
+                f"CompositeImplicitAutograd operators.")
+
+        with no_dispatch():
+            unwrapped_args = tree_map(unwrap, args)
+            unwrapped_kwargs = tree_map(unwrap, kwargs)
+            unwrapped_rs = func(*unwrapped_args, **unwrapped_kwargs)
+            rs = tree_map(wrap, unwrapped_rs)
+
+        if is_view_fn(func):
+            # Autograd asserts that for B = A.view_fn(...), B and A's storages
+            # are the same. Here we try to make B alias A to avoid those asserts.
+            aliased_tensor = args[0]
+            with no_dispatch():
+                x = torch.empty(aliased_tensor.shape, dtype=aliased_tensor.dtype,
+                                device=aliased_tensor.device)
+                x.set_(aliased_tensor)
+                args_with_x = list(args)
+                args_with_x[0] = x
+                result = func(*args_with_x, **kwargs)
+                if isinstance(result, tuple) or isinstance(result, list):
+                    for a, b in zip(rs, result):
+                        a.set_(b)
+                else:
+                    rs.set_(result)
+
+        # Some operations are allowed to in-place modify the metadata of the
+        # inputs. The only ones are the "inplace view functions"; when we
+        # run into these, we manually modify the metadata of the input.
+        with no_dispatch():
+            if is_inplace_view_fn(func):
+                func(args[0])
+
+        # For each CompositeCompliantTensor t, we check that t and t.elem
+        # have consistent metadata. If they don't have consistent metadata,
+        # that means the operator did something fishy.
+        check = partial(check_metadata_consistency)
+        tree_map(check, args)
+        tree_map(check, kwargs)
+        tree_map(check, rs)
+        return rs
+
+# The general strategy is to wrap all Tensor args and kwargs in
+# CompositeCompliantTensor wrappers. If an operator that is
+# CompositeImplicitAutograd does any non-compliant behavior,
+# CompositeCompliantTensor will raise an error.
+def _check_composite_compliance(op, args, kwargs):
+    def wrap(e):
+        return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
+
+    args = tree_map(wrap, args)
+    kwargs = tree_map(wrap, kwargs)
+    try:
+        with enable_python_mode(CompositeCompliantTensor):
```

Comment:
Why do you need this to be a mode?

Reply:
We would like to catch CompositeImplicitAutograd implementations where someone creates an empty tensor and then calls `.resize_` on it. The empty tensor is not an input of the function but the mode makes it so that it gets wrapped in the CompositeCompliantTensor